### PR TITLE
FIX Downsampling Bugs

### PR DESCRIPTION
**Pull request by xir4n**
Changed default arguments in DTCWT:
1. Set `alternate_gh=False` as the default value for `dtcwt`
2.  Resolved an issue where the normalization during downsampling was incorrect.
4. Enhanced to_conv1d method
5. Adjusted the downsampling level to optimize performance and accuracy.

### CODE REVIEW

File: tests/test_nn.py [hunk 1]
```diff
@@ -90,13 +91,16 @@ def test_modulus():
...
         Q=Q,
         T=T,
         in_channels=2,
     )
-    conv1d = tfm.to_conv1d()
-    assert isinstance(conv1d, torch.nn.Conv1d)
\ No newline at end of file
+    x = torch.zeros(1, 1, 2**J)
+    conv1ds = tfm.to_conv1d()
+    for k, v in conv1ds.items():
```
Comment (lostanlen) [inline tests/test_nn.py:104]: or, more simply, `for v in conv1ds.values()`

File: tests/test_nn.py [hunk 2]
```diff
@@ -90,13 +91,16 @@ def test_modulus():
...
+        assert isinstance(v, torch.nn.Conv1d)
+        y = v(x)
```
Comment (lostanlen) [inline tests/test_nn.py:106]: test something about y?

File: tests/test_dtcwt.py [hunk 3]
```diff
@@ -129,7 +131,7 @@ def test_avrg_energy(alternate_gh):
     P_phi = torch.linalg.norm(phi) ** 2 / phi.shape[-1]
     P_Ux = P_Ux + P_phi
     for psi in psis:
-        psi = psi / math.sqrt(2) #?
+        psi = psi / math.sqrt(2)
```
Comment (lostanlen) [inline tests/test_dtcwt.py:134]: :)

File: tests/test_dtcwt.py [hunk 4]
```diff
@@ -115,12 +116,13 @@ def test_energy_preservation(alternate_gh):
...
 
+@pytest.mark.parametrize("J", range(3))
 @pytest.mark.parametrize("alternate_gh", [True, False])
-def test_avrg_energy(alternate_gh):
+def test_avrg_energy(J, alternate_gh):
     '''
     Test the power of the signals for normalization case.
     '''
-    tfm = murenn.DTCWT(alternate_gh=alternate_gh, normalize=True)
+    tfm = murenn.DTCWT(J=J+1, alternate_gh=alternate_gh, normalize=True)
```
Comment (lostanlen) [inline tests/test_dtcwt.py:125]: i'd prefer if you had `J=J` in the callsite and `range(1, 4)` in the `pytest .mark` parametrization macro

File: tests/test_dtcwt.py [hunk 5]
```diff
@@ -81,12 +81,13 @@ def test_skip_hps(skip_hps, normalize):
...
 
+@pytest.mark.parametrize("J", range(3))
 @pytest.mark.parametrize("alternate_gh", [True, False])
-def test_phi(alternate_gh):
+def test_phi(J, alternate_gh):
     '''
     Test the low-pass output phi doesn't diverge.
     '''
-    tfm = murenn.DTCWT(alternate_gh=alternate_gh, include_scale=True, skip_hps=True)
+    tfm = murenn.DTCWT(J=J+1, alternate_gh=alternate_gh, include_scale=True, skip_hps=True)
```
Comment (lostanlen) [inline tests/test_dtcwt.py:90]: i'd prefer if you had `J=J` in the callsite and `range(1, 4)` in the `pytest .mark` parametrization macro

File: murenn/dtcwt/nn.py [hunk 6]
```diff
@@ -179,13 +191,13 @@ def __init__(self, J_phi):
...
             skip_hps=True,
-            padding_mode="zeros",
         )
+        self.relu = torch.nn.ReLU()
 
 
     def forward(self, x):
         for j in range(self.J_phi):
             x, _ = self.phi(x)
             # Normalize the coefficients
```
Comment (lostanlen) [inline murenn/dtcwt/nn.py:201]: remove this comment. you are not normalizing the coefficients anymore. you are directly applying a decimation `::2`

File: murenn/dtcwt/nn.py [hunk 7]
```diff
@@ -179,13 +191,13 @@ def __init__(self, J_phi):
...
-            x = x[:,:,::2] / math.sqrt(2)
-        return x
\ No newline at end of file
+            x = x[:,:,::2]
+        return self.relu(x)
```
Comment (lostanlen) [inline murenn/dtcwt/nn.py:203]: this line deserves a one-line comment above to justify the use of `relu`

File: murenn/dtcwt/nn.py [hunk 8]
```diff
@@ -96,44 +94,58 @@ def to_conv1d(self):
         """
```
Comment (lostanlen) [inline murenn/dtcwt/nn.py:94]: the return value docstring is outdated

File: murenn/dtcwt/nn.py [hunk 9]
```diff
@@ -179,13 +191,13 @@ def __init__(self, J_phi):
             J=1,
             level1="near_sym_b",
             skip_hps=True,
-            padding_mode="zeros",
```
Comment (lostanlen) [inline murenn/dtcwt/nn.py:182]: ?

File: murenn/dtcwt/nn.py [hunk 10]
```diff
@@ -88,52 +86,66 @@ def to_conv1d(self):
         Compute the single-resolution equivalent impulse response of the MuReNN layer.
         This would be helpful for visualization in Fourier domain, for receptive fields,
         and for comparing computational costs.
-           DTCWT        conv1d        IDTCWT
-        δ -------> ψ_j --------> w_jq -------> y_jq
+           conv1d        IDTCWT
+        δ --------> w_jq -------> y_jq
```
Comment (lostanlen) [inline murenn/dtcwt/nn.py:90]: i don't think this diagram is correct either. we don't start from a dirac in the time domain anymore. instead we build a dirac in the DTCWT domain, on either tree (a and b)

File: murenn/dtcwt/nn.py [hunk 11]
```diff
@@ -88,52 +86,66 @@ def to_conv1d(self):
...
         x = torch.zeros(1, self.in_channels, N).to(device)
-        x[:, :, N//2] = 1
-        inv = murenn.IDTCWT(
-            J = J,
-            alternate_gh=False         
-        ).to(device)
-        # Get DTCWT impulse reponses
+
+        # Initialize the inverse DTCWT
+        inv = murenn.IDTCWT(J=J, alternate_gh=False).to(device)
```
Comment (lostanlen) [inline murenn/dtcwt/nn.py:108]: what's the status of alternate_gh now? are you sure it's still necessary to change the default to `False`? what happens if you set `alternate_gh=True`?

File: murenn/dtcwt/nn.py [hunk 12]
```diff
@@ -88,52 +86,66 @@ def to_conv1d(self):
...
-            out_channels=ws.shape[0],
-            kernel_size=N,
-            bias=False,
-            padding="same",
-        )
-        conv1d.weight.data = torch.nn.parameter.Parameter(ws)
-        return conv1d
+                Wpsi_jq = Wpsi_j[0, q, :].reshape(1,1,-1)
+                Wpsis_r = [Wpsi_jq * (1+0j) if k == j else psis[k].new_zeros(1,1,psis[k].shape[-1]) for k in range(J)]
+                Wpsis_i = [Wpsi_jq * (0+1j) if k == j else psis[k].new_zeros(1,1,psis[k].shape[-1]) for k in range(J)]
```
Comment (lostanlen) [inline murenn/dtcwt/nn.py:125]: excellent!!

File: murenn/dtcwt/transform1d.py [hunk 13]
```diff
@@ -16,7 +16,7 @@ def __init__(
         J=8,
         skip_hps=False,
         include_scale=False,
-        alternate_gh=True,
+        alternate_gh=False,
```
Comment (lostanlen) [inline murenn/dtcwt/transform1d.py:19]: i'm still curious to know if this breaking change is still useful

File: tests/test_nn.py [hunk 14]
```diff
@@ -89,14 +91,39 @@ def test_modulus():
...
+    )
+    # Initialize the learnable filters with dirac
+    for conv1d in tfm.conv1d:
+        torch.nn.init.dirac_(conv1d.weight)
+    # Get the dtcwt filters
+    psis = tfm.to_conv1d()
+    # Test the energy crossing subbands
+    y = psis["complex"](torch.complex(x, x)/math.sqrt(2))
+    energy = torch.linalg.norm(y, dim=-1)
+    assert torch.allclose(energy, torch.ones(1, J), atol=0.1)
```
Comment (lostanlen) [inline tests/test_nn.py:129]: the new tests look great

**Comment by lostanlen** [comment]
Hello @xir4n , i have provided some more feedback. Please click the "Resolve conversation" to mark my comments as resolved. The two remaining things are:
- this diagram, which i don't think is correct
```
           conv1d        IDTCWT
        δ --------> w_jq -------> y_jq
```
- whether defaulting to `alternate_gh=False` is still necessary

Thanks!

File: murenn/dtcwt/nn.py (hunk 10 above)
Comment (xir4n) [reply murenn/dtcwt/nn.py:90]: how should I describe this operation? use $\delta_j$ instead of $\delta$?

File: murenn/dtcwt/nn.py (hunk 11 above)
Comment (xir4n) [reply murenn/dtcwt/nn.py:108]: Thanks for asking, this script shows the effect of `alternate_gh`:
```
from murenn import DTCWT, IDTCWT
import torch
from matplotlib import pyplot as plt
import numpy as np

def get_basis(j, **kwargs):
    N = 2**12
    x = torch.zeros(1, 1, N)

    fwd = DTCWT(**kwargs)
    inv = IDTCWT(**kwargs)

    lp, bpsr = fwd(x)
    _, bpsi = fwd(x)
    _, bpsc = fwd(x)


    bpsj = torch.zeros(bpsr[j].shape)
    bpsj[:,:,bpsj.shape[2]//2] = 1 / np.sqrt(2)**j

    bpsr[j] = bpsj * (1+0j)
    bpsi[j] = bpsj * (0+1j)


    yr = inv(lp, bpsr).detach()
    yi = inv(lp, bpsi).detach()

    return yr, yi

for j in range(8):
    yr, yi = get_basis(j, alternate_gh=False)
    yc = yr + 1j*yi
    yc_hat = np.fft.fft(yc.squeeze().numpy())
    plt.plot(np.abs(yc_hat)/np.abs(yc_hat).max())
plt.title("alternate_gh=False")

for j in range(8):
    yr, yi = get_basis(j, alternate_gh=True)
    yc = yr + 1j*yi
    yc_hat = np.fft.fft(yc.squeeze().numpy())
    plt.plot(np.abs(yc_hat)/np.abs(yc_hat).max())
plt.title("alternate_gh=True")
```

image
image

File: murenn/dtcwt/transform1d.py (hunk 13 above)
Comment (xir4n) [reply murenn/dtcwt/transform1d.py:19]: i think this is still useful

File: murenn/dtcwt/nn.py (hunk 10 above)
Comment (lostanlen) [reply murenn/dtcwt/nn.py:90]: re-reading the diagram, i now think it's actually good!
